// groovebasin/lib/plugin.js
var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

exports.Plugin = (function() {

  function _Class(log, onStateChanged, onStatusChanged) {
    this.log = log;
    this.onStateChanged = onStateChanged;
    this.onStatusChanged = onStatusChanged;
    this.onSendStatus = __bind(this.onSendStatus, this);
    this.onSocketConnection = __bind(this.onSocketConnection, this);
    this.setMpd = __bind(this.setMpd, this);
    this.setConf = __bind(this.setConf, this);
    this.handleRequest = __bind(this.handleRequest, this);
    this.restoreState = __bind(this.restoreState, this);
    this.saveState = __bind(this.saveState, this);
    this.mpd = null;
    this.conf = null;
    this.is_enabled = true;
  }

  _Class.prototype.saveState = function(state) {};

  _Class.prototype.restoreState = function(state) {};

  _Class.prototype.handleRequest = function(request, response) {
    return false;
  };

  _Class.prototype.setConf = function(conf, conf_path) {};

  _Class.prototype.setMpd = function(mpd) {};

  _Class.prototype.onSocketConnection = function(socket, getPermissions) {};

  _Class.prototype.onSendStatus = function(status) {};

  return _Class;

})();
